dex/networks/eth/bondcontracts/v0/contracts/ETHBondV0.sol

```diff
@@ -0,0 +1,262 @@
+// SPDX-License-Identifier: BlueOak-1.0.0
+// pragma should be as specific as possible to allow easier validation.
+pragma solidity = 0.8.18;
+
+// ETHBond is an ethereum contract used to manage fidelity bonds. It allows
+// users to bond funds until a certain locktime. When the locktime passes, the
+// bond can be refunded.
+// Bonds can also be updated. One or more bonds can be used to create a new bond
+// with a longer locktime. The new bond can have a higher value than all of the
+// bonds used to create it, requiring the user the send additional funds. The
+// new bond can also have a lower value than the bonds used to create it, in
+// which case a "change" bond is also created with a locktime equal to the
+// locktime of the earliest expiring bond used to create it.
+contract ETHBond {
+
+    // Bond holds information about a single bond.
+    struct Bond {
+        uint value;
+        uint initBlockNumber;
+        uint64 locktime;
+        address owner;
+
+        // prevBondID and nextBondID are used to create a doubly-linked list
+        // of bonds for an account. This is needed to be able to lookup all of
+        // the bonds for an account.
+        bytes32 prevBondID;
+        bytes32 nextBondID;
+    }
```

Comment:
This is a lot of data to store. What about a (completely untested) construction like this...

```Solidity
// SPDX-License-Identifier: BlueOak-1.0.0
pragma solidity = 0.8.19;

contract ETHBond {
    struct Bond {
        uint256 value;
        uint64 locktime;
    }

    mapping(bytes32 => bytes32) bondCommits;

    function hashBond(Bond calldata bond) public pure returns (bytes32) {
        return sha256(bytes.concat(bytes32(bond.value), bytes8(bond.locktime)));
    }

    // hashBonds creates a hash that commits to the sender and the provided bonds.
    function hashBonds(Bond[] calldata bonds) public view returns (bytes32) {
        bytes32 h;
        for (uint i = 0; i < bonds.length; i++) {
            if (i == 0) {
                h = sha256(bytes.concat(bytes20(msg.sender), hashBond(bonds[i])));
            } else {
                h = sha256(bytes.concat(h, hashBond(bonds[i])));
            }
        }
        return h;
    }

    function updateBonds(bytes32 acctID, Bond[] calldata oldBonds, Bond[] calldata newBonds) external payable {
        bytes32 commit = bondCommits[acctID];
        if (uint256(commit) != 0) {
            require(oldBonds.length > 0, "need old bonds");
            require(hashBonds(oldBonds) == commit, "wrong old bonds or sender");
        } else {
            require(oldBonds.length == 0, "old bonds provided but no commit is known");
        }
        // Quick path for pure refund.
        if (newBonds.length == 0) {
            uint256 refund = 0;
            for (uint i = 0; i < oldBonds.length; i++) {
                require(oldBonds[i].locktime < block.timestamp, "cannot refund unexpired bond");
                refund += oldBonds[i].value;
            }
            (bool ok, ) = payable(msg.sender).call{value: refund}("");
            require(ok == true, "full refund failed");
            bondCommits[acctID] = bytes32(0);
            return;
        }
        uint64 stamp = newBonds[newBonds.length - 1].locktime;
        uint256 newFunds = 0;
        Bond[] memory bondReserves = oldBonds;
        // Fund new bonds.
        for (uint i = newBonds.length - 1; i >= 0; i--) {
            Bond memory newBond = newBonds[i];
            require(newBond.locktime <= stamp, "bonds not sorted");
            stamp = newBond.locktime;
            uint256 toFund = newBond.value;
            for (uint j = bondReserves.length - 1; j >= 0; j--) {
                require(bondReserves[j].locktime < newBond.locktime, "illegal lock time shortening");
                if (bondReserves[j].value >= toFund) {
                    bondReserves[j].value -= toFund;
                    toFund = 0;
                    break;
                } else {
                    toFund -= bondReserves[j].value;
                    bondReserves[j].value = 0;
                }
            }
            newFunds += toFund;
        }
        // Are bonds not funded by extending old bonds funded by the transaction?
        require (msg.value >= newFunds, "unfunded bonds");
        // If some bonds weren't extended, refund them;
        uint256 unextended = 0;
        for (uint j = bondReserves.length - 1; j >= 0; j--) {
            if (bondReserves[j].value > 0) {
                require(bondReserves[j].locktime < block.timestamp, "unextended bond is not old enough to refund");
                unextended += bondReserves[j].value;
            }
        }
        if (unextended > 0) {
            (bool ok, ) = payable(msg.sender).call{value: unextended}("");
            require(ok == true, "unextended refund failed");
        }
        bondCommits[acctID] = hashBonds(newBonds);
        return;
    }

    function validateBonds(bytes32 acctID, Bond[] calldata bonds) external view returns (bool) {
        return bondCommits[acctID] == hashBonds(bonds);
    }
}
```

Reply:
If we were concerned about recovering bonds, It could also tweaked to something like.

```Solidity
    struct BondCommit {
        bytes32 commitHash;
        uint256 totalBonded;
        uint updateBlockNumber;
        address bonder;
        uint64 ultimateLocktime; // highest locktime of all bonds
    }

    mapping(bytes32 => BondCommit) bondCommits;
```

and then add recovery methods. This should still save a large amount on gas.

Reply:
This is good.

Reply:
With this code anyone could lock down an accountID that does not belong to them, if that is a concern.

Reply:
The user has the private key for the account id don't they? Seems they could sign a hash of the info they are sending and send the signature too. The contract could verify the signature and then the owner of the id could refund anywhere and such as only they could touch thier account id bonds.

Reply:
> With this code anyone could lock down an accountID that does not belong to them, if that is a concern.

Yeah this is not ideal. 

> The user has the private key for the account id don't they? Seems they could sign a hash of the info they are sending and send the signature too. The contract could verify the signature and then the owner of the id could refund anywhere and such as only they could touch thier account id bonds.

The account ID is the blake256 hash of the public key, and Ethereum does not support blake256, so I'm not sure how to solve this.

Reply:
Maybe just allow multiple commits per account ID.

Reply:
`hashBonds` commits to the `msg.sender`, so only the user who creates the bonds can update the bonds

Reply:
Oh. You're saying somebody could create bonds for an account ID that they don't own and prevent someone else from doing that. Mmmkay. The `mapping` key should commit to both account ID and address I guess.

Reply:
What about having a `mapping(bytes32 => bytes32[])`. Then each account ID could have multiple commits. Generally only the first one would be used though, unless someone's messing around.